fs_mol/utils/protonet_utils.py

```diff
@@ -75,6 +87,59 @@ def run_on_batches(
     return total_loss.cpu().item() / total_num_samples, metrics
 
 
+def validate_by_finetuning_on_tasks(
+    model: PrototypicalNetwork,
+    dataset: FSMolDataset,
+    seed: int = 0,
+    aml_run=None,
+    metric_to_use: MetricType = "avg_precision",
+) -> float:
+    """
+    Validation function for prototypical networks. Similar to test function;
+    each validation task is used to evaluate the model more than once, the
+    final results are a mean value for all tasks over the requested metric.
+    """
+
+    batcher = get_protonet_batcher(max_num_graphs=model.config.batch_size)
+
+    def test_model_fn(
+        task_sample: FSMolTaskSample, temp_out_folder: str, seed: int
+    ) -> BinaryEvalMetrics:
+        pn_task_sample = task_sample_to_pn_task_sample(task_sample, batcher)
+
+        _, result_metrics = run_on_batches(
+            model,
+            batches=pn_task_sample.batches,
+            batch_labels=pn_task_sample.batch_labels,
+            train=False,
+        )
+        logger.info(
+            f"{pn_task_sample.task_name}:"
+            f" {pn_task_sample.num_support_samples:3d} support samples,"
+            f" {pn_task_sample.num_query_samples:3d} query samples."
+            f" Avg. prec. {result_metrics.avg_precision:.5f}.",
+        )
+
+        return result_metrics
+
+    task_results = eval_model(
+        test_model_fn=test_model_fn,
+        dataset=dataset,
+        train_set_sample_sizes=model.config.validation_support_set_sizes,
+        test_size_or_ratio=model.config.validation_query_set_size,
+        fold=DataFold.VALIDATION,
+        num_samples=model.config.validation_num_samples,
+        seed=seed,
+    )
```

Comment:
This is almost exactly the same as the `test`  function in  `protonet_test.py`. Could we extract out the common stuff? For example, `test_model_fn` looks like an exact copy, so we could have that outside?

Reply:
yup, we can do that.

Reply:
Comment here to summarize discussion elsewhere-- need to factor out more general function that `test_model_fn` because of the need for model to be a variable in scope.